From 724f26d7f244832b60de97593e1181288ee099e9 Mon Sep 17 00:00:00 2001
From: Avery Felts <averyfelts@aol.com>
Date: Thu, 22 Oct 2026 13:55:12 -0600
Subject: [PATCH 34/59] server: Give BUILTIN\Users access to group-accessible
 files.

From WineForge/MetalSharp (Epic install directories).

(cherry picked from commit 724cd7ef6c8714864a2be488ed1ab003e7448ba1)
---
 server/file.c | 32 ++++++++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/server/file.c b/server/file.c
index 11f3b0ed6db..91712661bed 100644
--- a/server/file.c
+++ b/server/file.c
@@ -304,6 +304,12 @@ struct security_descriptor *mode_to_sd( mode_t mode, const struct sid *user, con
         (!(mode & S_IWUSR) && (mode & (S_IWGRP|S_IWOTH))) ||
         (!(mode & S_IXUSR) && (mode & (S_IXGRP|S_IXOTH))))
         dacl_size += sizeof(*ace) + sid_len( user );
+    if (mode & S_IRWXG)
+    {
+        dacl_size += sizeof(*ace) + sid_len( group );
+        if (!equal_sid( group, &builtin_users_sid ))
+            dacl_size += sizeof(*ace) + sid_len( &builtin_users_sid );
+    }
     if (mode & S_IRWXO) dacl_size += sizeof(*ace) + sid_len( &world_sid );
 
     sd = mem_alloc( sizeof(*sd) + sid_len( user ) + sid_len( group ) + dacl_size );
@@ -329,6 +335,12 @@ struct security_descriptor *mode_to_sd( mode_t mode, const struct sid *user, con
         (!(mode & S_IWUSR) && (mode & (S_IWGRP|S_IWOTH))) ||
         (!(mode & S_IXUSR) && (mode & (S_IXGRP|S_IXOTH))))
         dacl->count++;
+    if (mode & S_IRWXG)
+    {
+        dacl->count++;
+        if (!equal_sid( group, &builtin_users_sid ))
+            dacl->count++;
+    }
     if (mode & S_IRWXO) dacl->count++;
 
     flags = (mode & S_IFDIR) ? OBJECT_INHERIT_ACE | CONTAINER_INHERIT_ACE : 0;
@@ -356,6 +368,26 @@ struct security_descriptor *mode_to_sd( mode_t mode, const struct sid *user, con
             ace->mask |= FILE_GENERIC_WRITE | DELETE | FILE_DELETE_CHILD;
         ace->mask &= ~STANDARD_RIGHTS_ALL; /* never deny standard rights */
     }
+    if (mode & S_IRWXG)
+    {
+        /* Primary group ACE, plus Epic install-dir compatibility for BUILTIN\Users below. */
+        ace = set_ace( ace_next( ace ), group, ACCESS_ALLOWED_ACE_TYPE, flags, WRITE_DAC | WRITE_OWNER );
+        if (mode & S_IRGRP) ace->mask |= FILE_GENERIC_READ | FILE_GENERIC_EXECUTE;
+        if (mode & S_IWGRP) ace->mask |= FILE_GENERIC_WRITE | DELETE | FILE_DELETE_CHILD;
+
+        if (!equal_sid( group, &builtin_users_sid ))
+        {
+            /*
+             * MetalSharp launcher-compat: mirror group access through BUILTIN\Users.
+             * MetalSharp-Internal: launcher-compat/epic-launcher-v1.
+             * MetalSharp-Feature: launcher-compat/epic-install-builtin-users-acl-v1.
+             */
+            ace = set_ace( ace_next( ace ), &builtin_users_sid, ACCESS_ALLOWED_ACE_TYPE,
+                           flags, WRITE_DAC | WRITE_OWNER );
+            if (mode & S_IRGRP) ace->mask |= FILE_GENERIC_READ | FILE_GENERIC_EXECUTE;
+            if (mode & S_IWGRP) ace->mask |= FILE_GENERIC_WRITE | DELETE | FILE_DELETE_CHILD;
+        }
+    }
     if (mode & S_IRWXO)
     {
         /* appropriate access rights for Everyone */
-- 
2.54.0 (Apple Git-157)

